### ISSUE TITLE
Broken graphics and sound on Batman Caped Crusader title screen

### ISSUE THREAD

**Issue by shinra-electric** [dirkwhoffmann/vAmiga#335]
Batman - The Caped Crusader.adf

Using v0.9.6

The title screen loads ok initially (though with broken sound):
![Screenshot 2020-05-11 at 09 43 46](https://user-images.githubusercontent.com/50119606/81519756-7ceed780-936c-11ea-9a04-5793f5bf1854.png)

But then turns to this: 

![Screenshot 2020-05-11 at 09 43 52](https://user-images.githubusercontent.com/50119606/81519776-8a0bc680-936c-11ea-9457-f259e28bfefe.png)

**Comment by shinra-electric** [comment]
Just tested with FS-UAE. The sound works fine, though there is a slight graphics glitch where the hackers (Vision Factory & Beyonders) inserted their message: 

![Screenshot 2020-05-11 at 10 03 04](https://user-images.githubusercontent.com/50119606/81520560-e1ab3180-936e-11ea-91f8-231e965f34a2.png)


![Screenshot 2020-05-11 at 10 03 02](https://user-images.githubusercontent.com/50119606/81520556-dce67d80-936e-11ea-9f69-22279842a1be.png)

**Comment by dirkwhoffmann** [comment]
Next patient please… 👩‍🔬

Let's have a quick look at the playfield parameters first 🤒: 

<img width="980" alt="Bildschirmfoto 2020-05-11 um 12 24 00" src="https://user-images.githubusercontent.com/12561945/81554104-7e2cfe00-9386-11ea-8db8-a250455d4020.png">

5 Bitplanes, lores mode… nothing special here. Let’s do an X-ray then…

<img width="796" alt="Bildschirmfoto 2020-05-11 um 12 55 16" src="https://user-images.githubusercontent.com/12561945/81554216-b03e6000-9386-11ea-831a-34433ade3502.png">

Interesting. There is Copper activity at the bottom, but none at the top. I am wondering how the Copper list looks like 🤔. 

<img width="356" alt="Bildschirmfoto 2020-05-11 um 12 26 37" src="https://user-images.githubusercontent.com/12561945/81554278-c8ae7a80-9386-11ea-9752-34005949a615.png">
<img width="356" alt="Bildschirmfoto 2020-05-11 um 12 26 51" src="https://user-images.githubusercontent.com/12561945/81554289-cea45b80-9386-11ea-98e1-1731399873dd.png">

OK, I see. The game doesn’t reset the bitplane pointers at the top of the screen as most games do. It resets the pointers at the bottom. Maybe vAmiga has problems with bitplane DMA switched on during VBLANK. I think I never really testes that 😬.

Next step: Write test cases with bitplane DMA switched on during VBLANK

**Comment by dirkwhoffmann** [comment]
This is what I’ve found out so far:

Copper list processing is fine. The issue is related to the interrupt that is triggered by the Copper at the end of the Copper list. Inside the interrupt handler, the bitplane pointers are readjusted to draw to main part of the screen. The pointers that are set by the Copper itself (inside the Copper list) are only used to display the „Press F1 ..“ stuff.

In vAmiga, the interrupt handler (which is pretty complicated) often needs more than 1 frame to complete. If the interrupt handler doesn’t finish in time, the handler is not called in the next frame and the bitplane pointers are not updated. This causes the corruption.

Now the question is: What is the reason for the delay? Unfortunately, it can be anything. 

I need to think about a clever debug strategy at this point…

**Comment by dirkwhoffmann** [comment]
This game is weird. 

- During boot, it triggers an address error exception by jumping to an odd address. The address error also occurs in SAE, so I think it's supposed to occur. Anyway, address error handling is supported, so that's most likely not the problem here. 

- The graphics glitches vanish if vAmiga is run with the FastBlitter (Blitter level 0). They also disappear when run with the SlowBlitter in immediate execution mode (Blitter level 1). Based on this observation, my guess is that the game changes some Blitter registers while the Blitter is operating (in level 0 or 1, the  Blitter is immune to such changes). 

Penguin and Joker shouldn't feel too save. I'm on their trail now...

**Comment by mithrendal** [comment]
>my guess is that the game changes some Blitter registers while the Blitter is operating 

but.. but ... is this allowed ? I mean why should you do this ? Yes maybe ... some mysterious blitter side effects ? Maybe it activates the secret and mysterious "F" channel  ...? No one has ever discovered of it ... all documentation only describe A B C and destination channel D. Or did they had some virtual channel multiplexing in mind.... strange...

**Comment by dirkwhoffmann** [comment]
> but.. but ... is this allowed ?

You definitely risk to get a ticket from Blitter police, but legend has it that they went out of service soon after Commodore broke down. 

Here is a more detailed summary of the current results:

Blitter level 0: 👍
Blitter level 1: 👍
Blitter level 2: 🥴
Blitter level 2 with SLOW_BLT_DEBUG(*): 👍

 (*) Fast execution of the microinstruction code (timing similar to the FastBlitter)

Hmm, this must be a timing issue, but it doesn’t seem to be register related. I’ve guarded all Blitter registers, but none of them reports to be written while the Blitter is running 🤔. I have to admit that this Joker guy is smart...

**Comment by mithrendal** [comment]
Maybe Joker uses the blitter finished interrupt ?  That could be a source for a timing issue?

http://amigadev.elowar.com/read/ADCD_2.1/Hardware_Manual_guide/node0125.html

**Comment by dirkwhoffmann** [comment]
> Maybe Joker uses the blitter finished interrupt ? That could be a source for a timing issue?

No, the Blitter interrupt is for beginners. vAmiga can’t be tricked out so easily any more (I hope 🙄). 

I am pretty sure Joker is smarter than that. And he is not only smart, he's just as mean 😖. During boot, he asks me to hit the left mouse button twice, and he performs a different number of Blitter operations depending on the delay. This makes the boot sequence indeterministic. 

But I already tricked him out by adding a new debug option that holds down the left mouse button permanently 😎. I think he didn’t expect me to do this, because the game does boot well with the mouse button pressed all the time. Now, I am able to compare FastBlitter runs with SlowBlitter runs. 

Here is the result (FastBlitter to the left, SlowBlitter to the right):

<img width="1366" alt="Bildschirmfoto 2020-05-12 um 13 51 03" src="https://user-images.githubusercontent.com/12561945/81687563-63778980-9459-11ea-9ff8-f7cf821e122e.png">

Up to Blit 18 (out of 580 🥵) everything is identical. Then, there is a difference in the setup for Blit 19. The setup happens in the middle of two Blitter operations which is an unusual source for timing issues.

Here is my proposed fix 😎: 

    if (game == "Batman" && blitCount == 19) pokeBLTDATA(0xc05b10);

BTW, there is a rumor that Windows 2000 had such code in it to make SimCity work 😂.

**Comment by dirkwhoffmann** [comment]
I've stopped the CPU at the location where the write to BLTAPTL happens:

<img width="980" alt="Bildschirmfoto 2020-05-12 um 14 26 14" src="https://user-images.githubusercontent.com/12561945/81691242-ccaccc00-945c-11ea-9877-05ca2a0be788.png">

🧐 Hmmm, yes, there is definitely a smoking gun here. Right at FC5DE0. 

Stay tuned. Need to write some test cases...

**Comment by mithrendal** [comment]
Yes yes ask Moira for help. Joker just did not expect that you recreated all the components of the Amiga Computer by your self. These clean and self created components are an unplanned obstacle for him ...

**Comment by dirkwhoffmann** [comment]
I think I am coming close. The screen corruption is most likely caused by a race condition between the CPU and the Blitter (happening in the very last blit before the title screen appears):

```
[1158] (  2,199) 065C06  0 --B--- 4018 0000 Memory: POKE16<8>(2d6)
[1158] (  2,201) 065C0A  0 --B--- 4018 0000 Memory: peek16<1>(65c0a)
[1158] (  2,201) 065C0A  0 --B--- 4018 0000 Memory: POKE16<8>(2d8)
[1158] (  2,203) 065C0A  0 --B--- 4018 0000 Memory: POKE16<1>(33e)
[1158] (  2,203) 065C0A  0 --B--- 4018 0000 Memory: POKE16<8>(2da)
[1158] (  2,205) 065C0A  0 --B--- 4018 0000 Memory: POKE16<1>(340)
[1158] (  2,205) 065C0A  0 --B--- 4018 0000 Memory: POKE16<8>(2dc)
[1158] (  2,207) 065C0A  0 --B--- 4018 0000 Memory: peek16<1>(65c0c)
[1158] (  2,207) 065C0A  0 --B--- 4018 0000 Memory: POKE16<8>(2de)
[1158] (  2,209) 065C0E  0 --B--- 4018 0000 Memory: peek16<1>(65c0e)
```

(The CPU is indexed <1> and the Blitter is indexed <8>)

Think of it as a car race. In vAmiga, the CPU overtakes the Blitter, but on a real Amiga, the Blitter wins.

**Comment by mithrendal** [comment]
How can it run faster than on a real Amiga ? ... 
you checked the correct bus cycles with tonis CPU tester did'nt you ? 
you implemented the Blitter DMA superiority Blitter, that does steal cycles from the CPU. Did'nt you ?
What about the "blitter-nasty" bit ? Is it set ? Then the CPU should not get any cycle at all, right ? According to your log it is maybe not... So every 4th cycle is left for the CPU ....

Hmmm how can that be ?? This is really interesting ...

**Comment by dirkwhoffmann** [comment]
> Hmmm so how can that be a race condition ?? This is really interesting ...

I told you, Joker is mean 😖. I think he fiddles around with the Blitter flags in DMACON.

**Comment by dirkwhoffmann** [comment]
Fixed in v0.9.7
